scripts/metalk8s_cli/mixins/salt.py

```diff
@@ -0,0 +1,119 @@
+import json
+import shlex
+import subprocess
+
+import six
+
+from metalk8s_cli.exceptions import CommandError, CommandInitError
+
+
+class SaltCommandMixin(object):
+    def __init__(self, args):
+        super(SaltCommandMixin, self).__init__(args)
+
+    def run_salt_minion(
+        self, command, local=False, outputter=None, color=True,
+        saltenv=None, pillar=None,
+    ):
+        full_cmd = ['salt-call']
+
+        if local:
+            full_cmd.append('--local')
+        if outputter:
+            full_cmd.extend(['--out', outputter])
+        if color:
+            full_cmd.append('--force-color')
+
+        full_cmd.extend(self._build_args(command))
+
+        if saltenv is not None:
+            full_cmd.append('saltenv={}'.format(saltenv))
+        if pillar is not None:
+            full_cmd.append("pillar='{}'".format(json.dumps(pillar)))
+
+        try:
+            return subprocess.check_output(full_cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
+                'Failed to execute "{}" [retcode {}]:\n{}'.format(
+                    ' '.join(full_cmd), exc.returncode, exc.output
+                )
+            )
+
+    def run_salt_master(self, command):
+        # TODO: add cache for the salt-master container ID
+        full_cmd = [
+            'crictl', 'exec', self.get_salt_master_container()
+        ]
+        master_cmd = ['salt-run', '--force-color'] + self._build_args(command)
+        full_cmd.extend(master_cmd)
+
+        try:
+            return subprocess.check_output(
+                # FIXME: pillar args... need to escape?
+                ' '.join(full_cmd), shell=True,
```

Comment:
hum yeah, this looks dangerous, passing a string coming from non-controlled input (pillar) and let the shell interpret it, with no escaping.

Why not use a list of string here?
And if you absolutely want to use a string, then use `shlex.quote`.

Now, being Python 2 it's probably not available (otherwise it's not fun), but IIRC the implementation is not that complex (but still, would be better to use a list of string and keep `shell=False`).

Reply:
My bad, the `FIXME` comment was not explicit. The issue I was faced with when using a list of strings, the value `"pillar='{bootstrap_id: bootstrap}'"` is not well interpreted by Salt, and fails with some `invalid pillar` exception. Maybe it's not well escaped when sent to the execution context, so I'll give a try to `shlex.quote` on this specific argument.

Reply:
Weird, the whole point of using a list of string is to avoid the need for escaping (because the values are directly passed to the process).

This works¹ for me:

```python
pillar = {"update_archive_name": "new_archive_name.zip"}
cmd = ['salt', '*', 'state.highstate', 'pillar', json.dumps(pillar)]
subprocess.run(cmd)
```

¹: works as in `doesn't complain about the pillar`, but it fails with `No Top file or master_tops data matches found` because my local salt is not configured :stuck_out_tongue:

Reply:
I did not have issues with direct usage of `salt-call` either, the issue arose when sending to Salt Master through `crictl exec`... Did not test yet if the `shlex.quote` approach helps, so we'll see :)

Reply:
Ha, you may need to wrap the pillar between single quote then.

I think we're doing something like that in the expansion test `tests/install/steps/test_expansion.py`, maybe you can get an idea there.

Reply:
Huh, it seems the tests are doing it with `shell=True` as well (using `host.run(some_format_string)`) :/ and through `kubectl`. Tried switching to `kubectl` as well, used different escaping strategies, nothing works. I'll go back to `shell=True` as well here, actually `pillar` is not provided by the user, so I'd say it's not really _unsafe_ to use it within the CLI code.

Reply:
Yeah indeed, the pillar value is hardcoded here so it should be safe: let's go for that.
(but it's weird that we can't make it works without this)